polaris/dataset/_base.py

```diff
@@ -0,0 +1,403 @@
+                "A Zarr archive associated with a Polaris dataset has to be consolidated."
+            ) from error
+        return self._zarr_root
+
+    @computed_field
+    @property
+    def n_rows(self) -> int:
+        """The number of rows in the dataset."""
+        return len(self.rows)
+
+    @computed_field
+    @property
+    def n_columns(self) -> int:
+        """The number of columns in the dataset."""
+        return len(self.columns)
+
+    @property
+    @abc.abstractmethod
+    def rows(self) -> list:
+        """Return all row indices for the dataset"""
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def columns(self) -> list:
+        """Return all columns for the dataset"""
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def dtypes(self) -> dict[str, np.dtype]:
+        """Return the dtype for each of the columns for the dataset"""
+        raise NotImplementedError
+
+    def load_to_memory(self):
+        """
+        Load data from zarr files to memeory
+
+        Warning: Make sure the uncompressed dataset fits in-memory.
+            This method will load the **uncompressed** dataset into memory. Make
+            sure you actually have enough memory to store the dataset.
+        """
+        data = self.zarr_data
+
+        if not isinstance(data, zarr.Group):
+            raise TypeError(
+                "The dataset zarr_root is not a valid Zarr archive. "
+                "Did you call Dataset.load_to_memory() twice?"
+            )
+
+        # NOTE (cwognum): If the dataset fits in memory, the most performant is to use plain NumPy arrays.
+        # Even if we disable chunking and compression in Zarr.
+        # For more information, see https://github.com/zarr-developers/zarr-python/issues/1395
+        self._zarr_data = load_zarr_group_to_memory(data)
+
+    @abc.abstractmethod
+    def get_data(self, row: str | int, col: str, adapters: Optional[List[Adapter]] = None) -> np.ndarray:
+        """Since the dataset might contain pointers to external files, data retrieval is more complicated
+        than just indexing the `table` attribute. This method provides an end-point for seamlessly
+        accessing the underlying data.
+
+        Args:
+            row: The row index in the `Dataset.table` attribute
+            col: The column index in the `Dataset.table` attribute
+            adapters: The adapters to apply to the data before returning it.
+                If None, will use the default adapters specified for the dataset.
+
+        Returns:
+            A numpy array with the data at the specified indices. If the column is a pointer column,
+                the content of the referenced file is loaded to memory.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def upload_to_hub(
+        self, access: Optional[AccessType] = "private", owner: Union[HubOwner, str, None] = None
```

Comment:
The `owner` parameter is optional because it could also be set as a `Dataset` attribute. See e.g. [here](https://github.com/polaris-hub/polaris/blob/main/polaris/hub/client.py#L600).

Reply:
I see... It feels a little confusing to me and I'd rather we have one way of setting ownership. There's a scenario here where I'm passing in a different owner than is set in the instance, and then the model I have locally doesn't match what's been uploaded to the Hub.

Reply:
I'll update it to require explicitly setting the `Dataset.owner` attribute! :+1:

Reply:
Actually, that would affect the code examples on the Hub, which we have shown in various presentations and social posts. Removing the `owner` attribute from the Dataset class is also not an option because we use it in the `PolarisFileSystem` class. Finally, we could update the local `dataset.owner` attribute to match what was uploaded to the Hub, but that may be an equally confusing side-effect.

Since this becomes a more challenging issue than I initially thought and because this code wasn't introduced in this PR, I'm going to leave it as is for now. I did create https://github.com/polaris-hub/polaris/issues/192 to track this.